geo.py

```diff
@@ -83,39 +200,118 @@ def location_info(name, kind, placename_hints=None):
             loc["country"] = code
             coords = coords_for_country(code)
 
+    # Heimsálfa
+    elif kind == "continent":
+        # Get continent ISO code
+        loc["continent"] = CONTINENTS.get(name)
+
     # Götuheiti
     elif kind == "street":
         # All street names in BÍN are Icelandic
-        loc["country"] = ICELAND_ISOCODE
-        coords = coords_for_street_name(name, placename_hints=placename_hints)
+        if name not in ICE_STREETNAME_BLACKLIST:
+            loc["country"] = ICELAND_ISOCODE
+            coords = coords_for_street_name(name, placename_hints=placename_hints)
 
     # Örnefni
     elif kind == "placename":
-        info = icelandic_placename_info(name)
-        if info:
+        info = None
+        
+        # Check if it's an Icelandic placename
+        if name in ICE_REGIONS:
             loc["country"] = ICELAND_ISOCODE
-            # Pick first matching placename, w/o disambiguating
-            # TODO: This could be smarter
-            coords = coords_from_addr_info(info[0])
+        elif name not in ICE_PLACENAME_BLACKLIST:
+            # Try to find a matching Icelandic placename
+            info = placename_lookup(name)
+            if info:
+                loc["country"] = ICELAND_ISOCODE
+                # Pick first matching placename
+                coords = coords_from_addr_info(info[0])
+
+        # OK, not Icelandic. Let's see if it's a foreign city
+        if not info:
+            cities = lookup_city_info(name)
+            if cities:
+                # Pick first match. Cityloc package should give us a match list
+                # with capitals given precedence, and ordered by population.
+                c = cities[0]
+                loc["country"] = c.get("country")
+                coords = coords_from_addr_info(c)
+
+    # Look up continent code for country
+    if "country" in loc:
+        loc["continent"] = continent_for_country(loc["country"])
 
     if coords:
         (loc["latitude"], loc["longitude"]) = coords
 
     return loc
 
 
+ICE_CITY_NAMES = None
+ICE_CITIES_JSONPATH = "resources/cities_is.json"
+
+
+def _load_city_names():
+    """ Load data from JSON file mapping Icelandic city names
+        to their corresponding English/international name. """
+    global ICE_CITY_NAMES
+    if ICE_CITY_NAMES is None:
+        jsonstr = resource_stream(__name__, ICE_CITIES_JSONPATH).read().decode()
```

Comment:
Since Reynir is not installed via `pip`, it is probably not canonical to use `resource_stream` here. Good ol' fashioned `open()` (after `os.path.join()` to construct the file name) is probably better.